buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java

```diff
@@ -363,11 +363,18 @@ static void setBytes(AbstractByteBuf buf, long addr, int index, ByteBuffer src)
             // Copy from direct memory
             long srcAddress = PlatformDependent.directBufferAddress(src);
             PlatformDependent.copyMemory(srcAddress + src.position(), addr, src.remaining());
-        } else {
+        } else if (src.hasArray()) {
             // Copy from array
             PlatformDependent.copyMemory(src.array(), src.arrayOffset() + src.position(), addr, length);
+        } else {
+            byte[] buffer = new byte[length];
+            src.get(buffer, 0, length);
+            PlatformDependent.copyMemory(buffer, 0, addr, length);
+        }
+
+        if (!src.isReadOnly()) {
```

Comment:
hmm... I think this is not correct. This method should always adjust the position of the src buffer. Can you explain why you think this is correct ?